macros/src/main/scala/chisel3/internal/SubTypeOf.scala

```diff
@@ -0,0 +1,49 @@
+package chisel3.internal
+
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox.Context
+
+trait ChiselSubTypeOf[A, B]
+
+object ChiselSubTypeOf {
+  def genChiselSubTypeOf[A: c.WeakTypeTag, B: c.WeakTypeTag](c: Context): c.Tree = {
+    import c.universe._
+
+    val ta = implicitly[c.WeakTypeTag[A]]
+    val tb = implicitly[c.WeakTypeTag[B]]
+    val empty = q""
+
+    // Returns true if 'a' and 'b' are structurally equivalent types.
+    def typeEquals(a: Type, b: Type): Boolean = {
+      if (a == NoType || b == NoType) {
+        return false
+      }
+      // Two types are equal if they are both subtypes of each other.
+      subtypeOf(a, b) && subtypeOf(b, a)
+    }
+
+    // Returns true if 'a' is a structural subtype of 'b' (i.e. all fields of
+    // 'b' exist within 'a' with the same names and the same types).
+    def subtypeOf(a: Type, b: Type): Boolean = {
+      val mb = b.members.filter(_.isPublic)
+      // Go through every public member of b and make sure a member with the
+      // same name exists in a and it has the same structural type.
+      mb.filter(vb => vb.isTerm && vb.asTerm.isGetter || vb.isType)
+        .forall(vb => {
+          val name = if (vb.isTerm) TermName(vb.name.toString) else TypeName(vb.name.toString)
+          typeEquals(a.member(name).info, vb.info)
+        })
+    }
```

Comment:
This doesn't actually check that the members are of type `Data`, so I can make things not work by adding fields of other Scala types that should be irrelevant, eg. `val d = 3`. See [the plugin](https://github.com/chipsalliance/chisel3/blob/5fdf74f95e64cb69d6097547f20d789a83dbd735/plugin/src/main/scala/chisel3/internal/plugin/ChiselComponent.scala#L32) for a complex example of checking the types of things, we could probably do something similar (but cleaner) here.